tensorflow/core/kernels/ctc_loss_op.cc

```diff
@@ -25,8 +29,86 @@ limitations under the License.
 #include "tensorflow/core/util/ctc/ctc_loss_calculator.h"
 #include "tensorflow/core/util/sparse/sparse_tensor.h"
 
+#if GOOGLE_CUDA
+#include "tensorflow/core/platform/stream_executor.h"
+#include "tensorflow/core/util/stream_executor_util.h"
+#endif // GOOGLE_CUDA
+
 namespace tensorflow {
 
+typedef Eigen::ThreadPoolDevice CPUDevice;
+#if GOOGLE_CUDA
+using GPUDevice = Eigen::GpuDevice;
+
+namespace {
+using se::DeviceMemory;
+using se::Stream;
+using se::StreamExecutor;
+using se::ScratchAllocator;
+using se::dnn::CtcLossDescriptor;
+using se::dnn::RnnStateTensorDescriptor;
+using se::dnn::ToDataType;
+using se::port::StatusOr;
+
+template<typename T>
+void DoHistogram(OpKernelContext* ctx, const Tensor* labels_indices,
+                 int num_indices, int batch_size,
+                 std::vector<int> *labels_lengths) {
+  const T* h_in = labels_indices->flat<T>().data();
+  for(int i = 0; i < num_indices; i++) {
+    T key = h_in[i * 2];
+    (*labels_lengths)[key]++;
+  }
+}
+
+// A helper to allocate temporary scratch memory for cudnnCTCLoss ops. It
+// takes the ownership of the underlying memory. The expectation is that the
+// memory should be alive for the span of the cudnnCTCLoss itself.
+template <typename T>
+class CudnnCtcLossAllocatorInTemp : public ScratchAllocator {
```

Comment:
This code is identical to e.g. CudnnBatchNormAllocatorInTemp in fused_batch_norm_op.cc. Can you consolidate them to a single location instead of duplicating code, please?